### ISSUE TITLE
Global variables not accessible in listener `close` method

### ISSUE THREAD

**Issue by stasiek** [robotframework/robotframework#2443]
Opposed to Robot Framework 2.9.1, in Robot Framework 3.0 in the _close() method of a listener, the Robot "automatic variables" (e. g. ${OUTPUTDIR}) are not accessible anymore.

This is a problem, because you might want to use them during the cleanup actions in _close(). Looking at the "automatic variables" they seem to fall into different "scopes", i. e. some are only useful inside keywords, others in test cases, test suites or globally. In my opinion at least those of them which are useful globally should be available in _close().

an example
`tt.robot`

``` robotframework
*Settings*
Library  tt

*Test Case*
tc1
  Log  hello
```

`tt.py`

``` python
from robot.libraries.BuiltIn import BuiltIn

class MyListener(object):
    ROBOT_LISTENER_API_VERSION = 2

    def _close(self):
        print(BuiltIn().get_variable_value("${OUTPUT_DIR}"))

ROBOT_LIBRARY_LISTENER = MyListener()
```

Robot Framework 2.9.1:

```
% PYTHONPATH=. pybot tt.robot
==============================================================================
Tt                                                                           
==============================================================================
tc1                                                                   | PASS |
------------------------------------------------------------------------------
Tt                                                                    | PASS |
1 critical test, 1 passed, 0 failed
1 test total, 1 passed, 0 failed
==============================================================================
/tmp
Output:  /tmp/output.xml
Log:     /tmp/log.html
Report:  /tmp/report.html
```

Robot Framework 3.0:

```
% PYTHONPATH=. pybot tt.robot
==============================================================================
Tt                                                                           
==============================================================================
tc1                                                                   | PASS |
------------------------------------------------------------------------------
Tt                                                                    | PASS |
1 critical test, 1 passed, 0 failed
1 test total, 1 passed, 0 failed
==============================================================================
[ ERROR ] Unexpected error: RobotNotRunningError: Cannot access execution context
Traceback (most recent call last):
  File ".../local/lib/python2.7/site-packages/robot/utils/application.py", line 80, in _execute
    rc = self.main(arguments, **options)
  File ".../local/lib/python2.7/site-packages/robot/run.py", line 434, in main
    result = suite.run(settings)
  File ".../local/lib/python2.7/site-packages/robot/running/model.py", line 247, in run
    self.visit(runner)
  File ".../local/lib/python2.7/site-packages/robot/model/testsuite.py", line 160, in visit
    visitor.visit_suite(self)
  File ".../local/lib/python2.7/site-packages/robot/model/visitor.py", line 87, in visit_suite
    self.end_suite(suite)
  File ".../local/lib/python2.7/site-packages/robot/running/runner.py", line 108, in end_suite
    IMPORTER.close_global_library_listeners()
  File ".../local/lib/python2.7/site-packages/robot/running/importer.py", line 38, in close_global_library_listeners
    lib.close_global_listeners()
  File ".../local/lib/python2.7/site-packages/robot/running/testlibraries.py", line 198, in close_global_listeners
    listener._close()
  File "/tmp/tt.py", line 8, in _close
    print(BuiltIn().get_variable_value("${OUTPUT_DIR}"))
  File ".../local/lib/python2.7/site-packages/robot/libraries/BuiltIn.py", line 1114, in get_variable_value
    return self._variables[self._get_var_name(name)]
  File ".../local/lib/python2.7/site-packages/robot/libraries/BuiltIn.py", line 74, in _variables
    return self._namespace.variables
  File ".../local/lib/python2.7/site-packages/robot/libraries/BuiltIn.py", line 70, in _namespace
    return self._get_context().namespace
  File ".../local/lib/python2.7/site-packages/robot/libraries/BuiltIn.py", line 65, in _get_context
    raise RobotNotRunningError('Cannot access execution context')
```

**Comment by pekkaklarck** [comment]
Need to be investigated what change has caused this. Hopefully can be fixed in RF 3.0.1.

**Comment by haaja** [comment]
Just in case it helps I did some bisecting and I think commit 3b596411ea2d07e5f9c7d15df18876a16ad003d7 introduced this issue. It's quite large commit and I didn't have time to properly go through it. However, it seems that [end_suite](https://github.com/robotframework/robotframework/blob/4fda1dc/src/robot/running/runner.py#L92) function in runner.py ends test suite scope and variables before a possible call to close global listeners.

**Comment by pekkaklarck** [comment]
Investigated this a little more and it turned out that fixing this problem isn't exactly easy. Decided to move it into RF 3.1 scope for the following reasons:

- Required changes are too big and risky in a minor release.
- There haven't been more complaints about this problem.
- It is possible to access global variables in earlier listener methods and store values then. This ought to be an acceptable workaround for those who encounter this issue.

This issue also demonstrates a problem that errors in library listener close method crash the whole execution so that logs and reports aren't generated. This problem is easier to fix and I submitted #2509 about it already.

**Comment by pekkaklarck** [comment]
If the simple fix in PR #2626 really works (tests are needed!), this can be included in RF 3.0.3.

**Comment by pekkaklarck** [comment]
PR #2626 looks promising and nowadays also has a test. Testing could still be enhanced, but I'm fairly positive we can get this into RF 3.1 beta 1.
